docs/design/features/byreflike-generics.md

```diff
@@ -0,0 +1,75 @@
+# Generics parameters of ByRefLike types
+
+Using ByRefLike types in Generic parameters is possible by building upon support added for `ref` fields. Scenarios that would benefit most from this are those involving `Span<T>`. For example, consider the following examples:
+
+- `Span<TypedReference>` &ndash; Represents the general case where a ByRefLike type is used as a Generic parameter. This specific case would be desirable for a more efficient Reflection API.
+- `Span<Span<char>>` &ndash; Nested `Span<T>` types would be of benefit in the parsing result of strings.
+
+## Runtime impact
+
+Supporting ByRefLike type as Generic parameters will impact the following IL instructions:
+
+- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+- `stsfld` / `ldsfld` &ndash; Type fields of a ByRefLike parameter cannot be marked `static`.
+- `newarr` / `stelem` / `ldelem` / `ldelema` &ndash; Arrays are not able to contain ByRefLike types.
+    - `newobj` &ndash; For multi-dimensional array construction.
+- `constrained.callvirt` &ndash; This IL sequence must resolve to a method implemented on `object`, or a default interface method.
+
+If any of the above instructions are attempted to be used with a ByRefLike type, the runtime will throw an `InvalidProgramException`.
+
+The following instructions are already set up to support this feature since their behavior will fail as currently defined due to the inability to box a ByRefLike type.
+
+- `throw` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `unbox` / `unbox.any` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `isinst` &ndash; Will always place `null` on stack.
+- `castclass` &ndash; Will always throw `System.InvalidCastException`.
+
+## Proposal
+
+Support for the following would be indicated by the existing `RuntimeFeature.ByRefFields` mechanism.
```

Comment:
I believe that we are going to finish `ByRefFields` in .NET 7. This is going to be preview feature at best in .NET 7. I think this feature needs to get a new flag under `RuntimeFeature`.

Reply:
Are we certain the runtime side of this can't be finished in .NET 7? At present there do seem to be a fair number of changes, but I'm not seeing a specific red-flag that indicates this can't be completed. I agree that if something isn't in prior to the last preview before our first RC we aren't going to make.

Reply:
We can declare this finished only after both runtime and Roslyn sides are done, the end-to-end have seen some use and the feedback has been incorporated. If we want to ship it before that, it should be marked as preview. We even have annotations to mark it as such now.

Reply:
> We can declare this finished only after both runtime and Roslyn sides are done

That wasn't how I was defining this, but that does make it clear this won't be complete in .NET 7. What is the process for marking this in preview?

Reply:
Design doc: https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md

This is what similar situation looked like in .NET 6 (snippet https://github.com/dotnet/runtime/blob/release/6.0/src/libraries/System.Runtime/ref/System.Runtime.cs):
```
    public static partial class RuntimeFeature
    {
        [System.Runtime.Versioning.RequiresPreviewFeaturesAttribute("Generic Math is in preview.", Url = "https://aka.ms/dotnet-warnings/generic-math-preview")]
        public const string VirtualStaticsInInterfaces = "VirtualStaticsInInterfaces";
    }
```